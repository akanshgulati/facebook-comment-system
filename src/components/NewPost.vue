<template>
  <div :class="{'new-post--primary': isPrimary, 'new-post--secondary': !isPrimary}" class="new-post"
       @keydown.enter.exact.prevent="onSend">

    <UserIcon :url="currentUserIcon" class="new-post__img" :size="iconSize"></UserIcon>

    <div class="new-post__content">
      <textarea v-model="content" :placeholder="placeholder" rows="1" @input="onInput"></textarea>
    </div>
    <div class="new-post__control-buttons">
      <div class="new-post__control-buttons__send" @click="onSend">Send</div>
    </div>
  </div>
</template>

<script>
import UserIcon from './UserIcon';

export default {
  data() {
    return {
      content: ''
    };
  },
  methods: {
    onSend() {
      if (this.content.length) {
        // sending data
        this.$emit('onSend', {
          content: this.content,
          postedOn: +new Date(),
        });

        this.content = '';
        this.adjustTextAreaSize();
      }
    },
    adjustTextAreaSize() {
      if (!this.isPrimary) {
        return;
      }
      const textArea = this.$el.querySelector('textarea');
      textArea.style.fontSize = textArea.textLength > 0 ? (textArea.textLength > 80 ? '15px' : '20px') : '15px;';
      textArea.style.height = '';
      textArea.style.height = `${textArea.scrollHeight}px`;
    },
    onInput() {
      this.adjustTextAreaSize();
    },
    focusInput() {
      const textArea = this.$el.querySelector('textarea');
      textArea.focus();
    },
  },
  components: {
    UserIcon,
  },
  watch: {
    initContent(data) {
      this.content = data.content;
      this.focusInput();
    },
  },
  props: {
    placeholder: {
      type: String,
      default: 'Write a comment',
    },
    initContent: {
      type: Object,
    },
    isPrimary: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    iconSize() {
      return this.isPrimary ? 'medium' : 'small';
    },
    currentUserIcon() {
      return this.$store.state.currentUser.image;
    }
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>

  .new-post {
    border: 1px solid #e0e4ea;
    border-radius: 10px;
    display: flex;
    align-items: center;
    background-color: #fff;
  }

  .new-post--primary {
    padding: 10px 20px;
  }

  .new-post--secondary {
    padding: 5px 10px;
  }

  .new-post__img {
    align-self: flex-start;
  }

  .new-post__content {
    flex-grow: 1;
    padding: 0 20px;
    display: flex;
    align-items: center;
  }

  .new-post__content textarea {
    width: 100%;
    height: 100%;
    font-size: 15px;
    outline: 0;
    border: none;
    resize: none;
  }

  .new-post__control-buttons {
    flex-basis: auto;
  }
</style>
