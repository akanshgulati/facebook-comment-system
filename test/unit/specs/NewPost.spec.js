import Vue from 'vue';
import Vuex from 'vuex';
import NewPost from '@/components/NewPost';
import { createRenderer } from 'vue-server-renderer';

Vue.use(Vuex);

describe('NewPost.vue', () => {
  let store;
  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        currentUser: {
          image: '1.png',
        },
      },
    });
  });

  it('should match the snapshot', () => {
    const renderer = createRenderer();
    const Constructor = Vue.extend(NewPost);
    const vm = new Constructor({
      propsData: {
        placeholder: 'this is placeholder',
      },
      store,
    }).$mount();
    renderer.renderToString(vm, (err, str) => {
      expect(str).toMatchSnapshot();
    });
  });

  it('should match the snapshot with primary', () => {
    const renderer = createRenderer();
    const Constructor = Vue.extend(NewPost);
    const vm = new Constructor({
      propsData: {
        placeholder: 'this is placeholder',
        isPrimary: true,
      },
      store,
    }).$mount();
    renderer.renderToString(vm, (err, str) => {
      expect(str).toMatchSnapshot();
    });
  });

  it('should add correct user image', () => {
    const Constructor = Vue.extend(NewPost);
    const vm = new Constructor({
      propsData: {
        placeholder: 'this is placeholder',
        count: 2,
      },
      store,
    }).$mount();
    expect(vm.$el.querySelector('.new-post__img').hasAttribute('src')).toBeTruthy();
    expect(vm.$el.querySelector('.new-post__img').getAttribute('src')).toBe('1.png');
  });

  it('should add correct placeholder', () => {
    const Constructor = Vue.extend(NewPost);
    const vm = new Constructor({
      propsData: {
        placeholder: 'this is placeholder',
        count: 2,
      },
      store,
    }).$mount();
    const placeholderText = vm.$el.querySelector('.new-post__content').children[0].getAttribute('placeholder');
    expect(placeholderText).toBe('this is placeholder');
  });
});
