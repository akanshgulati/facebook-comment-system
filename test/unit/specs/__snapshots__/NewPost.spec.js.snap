// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPost.vue should match the snapshot 1`] = `
<div class="new-post new-post--secondary"><img src="1.png" alt="" class="user-icon new-post__img user-icon--small">
    <div class="new-post__content"><textarea placeholder="this is placeholder" rows="1"></textarea></div>
    <div class="new-post__control-buttons">
        <div class="new-post__control-buttons__send">Send</div>
    </div>
</div>
`;

exports[`NewPost.vue should match the snapshot with primary 1`] = `
<div class="new-post new-post--primary"><img src="1.png" alt="" class="user-icon new-post__img user-icon--medium">
    <div class="new-post__content"><textarea placeholder="this is placeholder" rows="1"></textarea></div>
    <div class="new-post__control-buttons">
        <div class="new-post__control-buttons__send">Send</div>
    </div>
</div>
`;
